### PR TITLE
Sockets - Pauline & Niv

### PR DESCRIPTION
**Pull request by paulentine**
# slack.rb
Congratulations! You're submitting your assignment!

You and your partner should _collaborate_ on the answers to these questions.

## Comprehension Questions
Question | Answer
---      | ---
How did you go about exploring the Slack API? Did you learn anything that would be useful for your next project involving an API? | We read the documentations, learned how to post request and how to conceal our token with VCR.
Give a short summary of the _request/response cycle_. Where does your program fit into that scheme? | The request/response cycle explains how an API retrieves information from the web. First, a request is made using HTTParty. Then, a response is sent back and we parse it for the information we want. Our program initiates this cycle when it retrieves the list of users and list of channels from slack and sends a message. 
How does your program check for and handle errors when using the Slack API? | We tests for error messages by posting bad requests, which should then raise a SlackApiError.
Did you need to make any changes to the design work we did in class? If so, what were they? | We chose to load our user and channels in self.load in each of these methods instead of using self.get(url, params). We loaded the lists into workspace(users, channels:) so that workspace would be less coupled to the class User or class Channel.
Did you use any of the _inheritance idioms_ we've talked about in class? How? | We set up template methods (self.load, details, self.list)  and polymorphism (send_message) in our abstract class (Recipient)
How does VCR aid in testing a program that uses an API? | VCR records an API request and its response so that when a test is called, it does not make a new call, but instead plays back the recorded response. This saves us from having to make multiple calls which can be costly and is useful if Slack's server is down.

### CODE REVIEW

File: lib/channel.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    def self.list
+      return @@all
+    end
+
+    def self.load
+      query_parameters = {
+        token: TOKEN,
+      }
+      response = HTTParty.get(URL, query: query_parameters)['channels']
+      response.each do |channel|
```
Comment (kaidamasaki) [inline lib/channel.rb:38]: Forgot to validate `response.code` and `response.parsed_response["ok"]`.

File: lib/recipient.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+      response = HTTParty.post("https://slack.com/api/chat.postMessage",
+                               body: body,
+                               headers: { "Content-Type" => "application/x-www-form-urlencoded" })
+
+      unless response.code == 200 && response.parsed_response["ok"]
+        raise SlackApiError, "Error when posting #{text} to #{name}, error: #{response.parsed_response["error"]}"
+      end
+
+      return response.code == 200 && response.parsed_response["ok"]
```
Comment (kaidamasaki) [inline lib/recipient.rb:26]: I would consider just returning `true` here since you've already done the check above.

File: lib/channel.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    def details
+    return "Channel name: #{name}
+     Slack ID: #{slack_id}
+     Topic: #{topic}
+     Member count: #{member_count}"
+    end
+
+    def self.list
+      return @@all
```
Comment (kaidamasaki) [inline lib/channel.rb:30]: While it doesn't cause a problem here generally class variables should be your tool of last resort.  I probably would have only stored them in a field in `Workspace` instead of here and there.

File: lib/user.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    end
+
+    def self.list
+      return @@all
+    end
+    
+    def self.load
+      query_parameters = { token: TOKEN }
+      response = HTTParty.get(BASE_URL << USERS_LIST_PATH, query: query_parameters)
+      response["members"].length.times do |i|
```
Comment (kaidamasaki) [inline lib/user.rb:39]: Forgot to validate `response.code` and `response.parsed_response["ok"]`.

File: lib/workspace.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  attr_reader :users, :channels
+  attr_accessor :selected
+
+    def initialize(users:, channels:, selected: nil)
+      @users = users
+      @channels = channels
+      @selected = selected
+    end
+
+    def select_channel(id_or_name:)
```
Comment (kaidamasaki) [inline lib/workspace.rb:18]: Tiny issue: it's a little odd to use a keyword argument here since you only have a single argument.

File: specs/channel_spec.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+    it "sends a message to selected channel" do
+      VCR.use_cassette("channel_send_message") do
+        channel = SlackAPI::Channel.list[0]
+        test2 = channel.send_message(text: "hello")
+        expect(test2).must_equal true
+       end
+    end
+
```
Comment (kaidamasaki) [inline specs/channel_spec.rb:39]: I'd like to see some edge case tests here.

What happens if you can't talk to the API or there's an error?

File: specs/recipient_spec.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
+require 'minitest'
+require_relative 'test_helper.rb'
+
+describe "Recipient" do
```
Comment (kaidamasaki) [inline specs/recipient_spec.rb:4]: Empty tests 🙁.

File: specs/recipient_spec.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+  before do
+
+  end
+
+  describe "self.get" do
+    
+  end
+
+  describe "details" do
+
```
Comment (kaidamasaki) [inline specs/recipient_spec.rb:14]: I would like to see an assertion that this `raise`s `NotImplementedError`.

File: specs/recipient_spec.rb [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  describe "self.list" do
+
```
Comment (kaidamasaki) [inline specs/recipient_spec.rb:18]: Same here.

File: lib/workspace.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    def select_user(id_or_name:)
+      @users.each do |user|
+        if user.name == id_or_name
+          @selected = user
+        elsif user.slack_id == id_or_name
+          @selected = user
+        end
+      end
+      return @selected
```
Comment (kaidamasaki) [inline lib/workspace.rb:37]: Ruby's `.find` enumerable method might be helpful to clean up this code.

File: specs/recipient_spec.rb [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  describe "details" do
+
+  end
+
+  describe "self.list" do
+    
+  end
+end
```
Comment (kaidamasaki) [inline specs/recipient_spec.rb:20]: I'd also like to see a test that verifies that `self.load` `raise`s `NotImplementedError`.

File: specs/user_spec.rb [hunk 12]
```diff
@@ -0,0 +1,39 @@
+require 'minitest'
+require_relative 'test_helper.rb'
+
+describe "User" do
+  before do 
+    VCR.use_cassette("load_users") do
+      @response = SlackAPI::User.load
```
Comment (kaidamasaki) [inline specs/user_spec.rb:7]: I'd like to see some edge case testing on `User.load`.

File: specs/user_spec.rb [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+      end
+  end
+
+  describe "self.list" do
+    it "prints out a table of users" do
+          expect(SlackAPI::User.list).must_be_kind_of Array
+    end
+  end
+
+  describe "send message" do
```
Comment (kaidamasaki) [inline specs/user_spec.rb:27]: I'd also like to see some edge case tests on sending messages.

File: specs/workspace_spec.rb [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+describe "Workspace" do
+  before do
+    VCR.use_cassette("load_workspace") do
+      response = SlackAPI::User.load
+      response2 = SlackAPI::Channel.load
+      @workspace = SlackAPI::Workspace.new(users: response, channels: response2)
+     end
+  end
+
+  describe "select_user" do
```
Comment (kaidamasaki) [inline specs/workspace_spec.rb:13]: I'd like a test on what happens if the user doesn't exist.  Also, do you lose your selection if you attempt to select an invalid user?  Should you?

**Comment by kaidamasaki** [comment]
# slack.rb
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene (no slack tokens) | 👍🏼
Comprehension questions | 👍🏼
**Functionality** |
List users/channels | 👍🏼
Select user/channel | 👍🏼
Show details | 👍🏼
Send message | 👍🏼
Program runs without crashing | 👍🏼
**Implementation** |
API errors are handled appropriately | Mostly.  (You forgot in a couple of places.)
Inheritance model matches in-class activity | Mostly.  (Changes were clearly outlined and well thought out.)
Inheritance idioms (abstract class, template methods, polymorphism) used appropriately | Yes.
Methods are used to break work down into simpler tasks | Yes.
Class and instance methods are used appropriately | Yes.
**Overall** | Great job!  Even though you missed a few edge cases your solution was really clear and well thought out.

**Comment by kaidamasaki** [comment]
# slack.rb
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene (no slack tokens) | 👍🏼
Comprehension questions | 👍🏼
**Functionality** |
List users/channels | 👍🏼
Select user/channel | `select user` works; `select channel` crashes.  It looks like you added an argument to `get_user_input` and forgot to pass it in.
Show details | 👍🏼
Send message | 👍🏼
Program runs without crashing | No, see above.
**Implementation** |
API errors are handled appropriately | Yes.
Inheritance model matches in-class activity | Yes.
Inheritance idioms (abstract class, template methods, polymorphism) used appropriately | Yes.
Methods are used to break work down into simpler tasks | Yes.
Class and instance methods are used appropriately | Yes.
**Overall** | Great job!  Aside from the `ArgumentError` mentioned above everything looks good.  Your code is clear, concise and well factored.
